### PR TITLE
New usb protocol for panda

### PR DESCRIPTION
**Pull request by briskspirit**
New panda usb protocol to support dynamic messages from 5 to 72 bytes long

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -121,6 +121,7 @@ tqdm = "*"
 urllib3 = "*"
 utm = "*"
 websocket_client = "*"
+bitstruct = "*"
```
Comment (adeebshihadeh) [inline Pipfile:124]: usually we add dependencies in their own PR

**Comment by briskspirit** [comment]
Test preserved route with uploaded rlogs ade3fc448c38401f|2021-11-07--15-02-02

**Comment by adeebshihadeh** [comment]
Can you make before and after timing histograms [like this](https://github.com/commaai/openpilot/pull/22402#issuecomment-953800915)?

**Comment by briskspirit** [comment]
image

File: selfdrive/boardd/panda.cc [hunk 2]
```diff
@@ -343,74 +344,121 @@ void Panda::send_heartbeat() {
...
-  uint32_t data[RECV_SIZE/4];
-  int recv = usb_bulk_read(0x81, (unsigned char*)data, RECV_SIZE);
+  uint8_t data[RECV_SIZE];
+  int recv = usb_bulk_read(0x81, (uint8_t*)data, RECV_SIZE);
 
   // Not sure if this can happen
   if (recv < 0) recv = 0;
 
-  if (recv == RECV_SIZE) {
+  if (recv == RECV_SIZE) { // TODO: Might change from full to overloaded? if > some threshold that is lower than RECV_SIZE, let's say 80-90%
```
Comment (adeebshihadeh) [inline selfdrive/boardd/panda.cc:411]:
```suggestion
  // TODO: Might change from full to overloaded? if > some threshold that is lower than RECV_SIZE, let's say 80-90%
  if (recv == RECV_SIZE) {
```

File: selfdrive/boardd/panda.cc [hunk 3]
```diff
@@ -343,74 +344,121 @@ void Panda::send_heartbeat() {
...
+      auto can_data = cmsg.getDat();
+      uint8_t data_len_code = len_to_dlc(can_data.size());
+      assert(can_data.size() <= (hw_type == cereal::PandaState::PandaType::RED_PANDA) ? 64 : 8);
+      assert(can_data.size() == dlc_to_len[data_len_code]);
+
+      if (cmsg.getAddress() >= 0x800) { // extended
+        *(uint32_t*)&send[pos+1] = (cmsg.getAddress() << 3) | (1 << 2);
+      } else { // normal
+        *(uint32_t*)&send[pos+1] = (cmsg.getAddress() << 3);
+      }
```
Comment (adeebshihadeh) [inline selfdrive/boardd/panda.cc:383]:
```suggestion
      *(uint32_t*)&send[pos+1] = (cmsg.getAddress() << 3);
      if (cmsg.getAddress() >= 0x800) { // extended
        *(uint32_t*)&send[pos+1] |= (1 << 2);
      }
```

**Comment by adeebshihadeh** [comment]
Is that the before or after?

**Pull request merged**